dronecan/driver/file.py

```diff
@@ -0,0 +1,158 @@
+            self.curr_frame = None
+            self.start_monotonic_ts = time.monotonic()
+            self.set_first_and_last_timestamp()
+        else:
+            self.readonly = False
+            self.file = open(filename, 'wb')
+            self.file.seek(0, os.SEEK_END)
+        self.first_receive = False
+
+    def close(self):
+        if self.file is not None:
+            self.file.close()
+
+    def __del__(self):
+        self.close()
+
+    def _read_frame(self):
+        # read header
+        header = self.file.read(20)
+        if len(header) < 20:
+            return None, None
+        magic, timestamp, crc, length, flags, msgid = struct.unpack('<HQHHHL', header)
+        timestamp = timestamp/1e6
+        if magic != FILE_MAGIC:
+            raise DriverError("invalid magic")
+        # read data
+        data = self.file.read(length)
+        if len(data) < length:
+            raise DriverError("short read")
+        # calculate CRC
+        crc2 = common.crc16_from_bytes(data)
+        if crc != crc2:
+            raise DriverError("CRC error")
+        # create frame
+        is_extended = (msgid & (1<<31)) != 0
+        is_canfd = flags & FILE_FLAG_CANFD != 0
+        canid = msgid & 0x1FFFFFFF
+        ts_monotonic = self.start_monotonic_ts + timestamp - self.start_file_monotonic_ts
+        frame = CANFrame(canid,data,is_extended,ts_monotonic,canfd=is_canfd)
+        return frame, timestamp
+
+    def set_first_and_last_timestamp(self):
+        # seek to beginning of file
+        self.file.seek(0, os.SEEK_SET)
+        # read first frame
+        frame, timestamp = self._read_frame()
+        if frame is None:
+            raise DriverError("empty file")
+        self.start_file_monotonic_ts = timestamp
+
+        # run through file to find last frame
+        while True:
+            frame, timestamp = self._read_frame()
+            self.end_file_monotonic_ts = timestamp
+            if frame is None:
+                break
+        # seek to start of file
+        self.file.seek(0, os.SEEK_SET)
+
+    def get_start_timestamp(self):
+        return self.start_file_monotonic_ts
+
+    def get_last_timestamp(self):
+        return self.end_file_monotonic_ts
+
+    def set_start_timestamp(self, timestamp):
+        self.start_monotonic_ts = timestamp
+
+    def receive(self, timeout=None):
+        if not self.readonly:
+            return None
+        if self.start_monotonic_ts is None:
+            self.start_monotonic_ts = time.monotonic()
+        if self.curr_frame is None:
+            self.curr_frame, _ = self._read_frame()
+            if self.curr_frame is None:
+                return None
+        if self.curr_frame.ts_monotonic > time.monotonic():
+            if timeout is not None and (self.curr_frame.ts_monotonic - time.monotonic()) <= timeout:
+                # sleep until timeout
+                time.sleep(self.curr_frame.ts_monotonic - time.monotonic())
```

Comment:
that could end up sleeping for a -ve time
I also think the timeout should be against the time the receive() call started, not the frame time, as it is now the sleep could end up being a lot longer than the passed in timeout